sdk/include/platform/sunburst/platform-ethernet.hh

```diff
@@ -0,0 +1,633 @@
+		return 0;
+	}
+
+	/**
+	 * Simple class representing a received Ethernet frame.
+	 */
+	struct Frame
+	{
+		uint16_t length;
+		uint8_t *buffer;
+
+		Frame(uint16_t length) : length(length)
+		{
+			buffer = new uint8_t[length];
+		}
+
+		Frame(const Frame &) = delete;
+		Frame(Frame &&other) : buffer(other.buffer), length(other.length)
+		{
+			other.buffer = nullptr;
+			other.length = 0;
+		}
+
+		~Frame()
+		{
+			if (buffer != nullptr)
+			{
+				delete[] buffer;
+			}
+		}
+	};
+
+	/**
+	 * Check the link status of the PHY.
+	 */
+	bool phy_link_status()
+	{
+		uint16_t status = register_read(RegisterOffset::P1SR);
+		return (status & Port1Status::LinkGood) != 0;
+	}
+
+	std::optional<Frame> receive_frame()
+	{
+		if (frames_to_process == 0)
+		{
+			uint16_t isr = register_read(RegisterOffset::ISR);
+			if (!(isr & ReceiveInterrupt))
+			{
+				return std::nullopt;
+			}
+
+			// Acknowledge the interrupt
+			register_write(RegisterOffset::ISR, ReceiveInterrupt);
+
+			// Read number of frames pending.
+			// Note that this is only updated when we acknowledge the interrupt.
+			frames_to_process = register_read(RegisterOffset::RXFCTR) >> 8;
+		}
+
+		// Get number of frames pending
+		for (; frames_to_process; frames_to_process--)
+		{
+			uint16_t status = register_read(RegisterOffset::RXFHSR);
+			uint16_t length = register_read(RegisterOffset::RXFHBCR) & 0xFFF;
+			bool     valid =
+			  (status & RxFrameValid) &&
+			  !(status &
+			    (RxCrcError | RxRuntFrame | RxFrameTooLong | RxMiiError |
+			     RxUdpFrameChecksumStatus | RxTcpFrameChecksumStatus |
+			     RxIpFrameChecksumStatus | RxIcmpFrameChecksumStatus));
+
+			if (!valid && DebugDroppedFrames)
+			{
+				if (DebugDroppedFrames)
+				{
+					Debug::log("Dropping frame with status: {}", status);
+				}
+
+				drop_error_frame();
+				continue;
+			}
+
+			if (length == 0)
+			{
+				if (DebugDroppedFrames)
+				{
+					Debug::log("Dropping frame with zero length");
+				}
+
+				drop_error_frame();
+				continue;
+			}
+
+			Debug::log("Receiving frame of length {}", length);
+
+			auto frame = Frame(length);
+
+			// Reset QMU RXQ frame pointer to zero and start DMA transfer
+			// operation.
+			register_write(RegisterOffset::RXFDPR, 0x4000);
+			register_set(RegisterOffset::RXQCR, StartDmaAccess);
+
+			// Start receiving via SPI.
+			uint8_t cmd = 0b10 << 6;
+			set_gpio_output_bit(EthCsPin, 0);
+			spi()->blocking_write(&cmd, 1);
+
+			// Initial words are RXFHSR and RXFHBCR which we have already know
+			// the value.
+			uint8_t dummy[8];
+			spi()->blocking_read(dummy, 8);
+
+			spi()->blocking_read(frame.buffer, length);
+
+			// Receive needs to be dword-aligned, read the paddings.
+			uint16_t pad = (-length) & 0x3;
+			if (pad != 0)
+			{
+				spi()->blocking_read(dummy, pad);
+			}
+
+			set_gpio_output_bit(EthCsPin, 1);
+
+			register_clear(RegisterOffset::RXQCR, StartDmaAccess);
+			frames_to_process -= 1;
+
+			return std::move(frame);
```

Comment:
The Arty A7 needs to cap permission because the buffer for frames are reused. For this driver I think it's actually fine even if receiver mutates the frame, so it doesn't really matter?

Reply:
Since now I have switched to a single buffer, I need to introduce the permission capping. The Arty A7 driver restrict permission on boundedBuffer but it doesn't seem to be used at all?